httpd-client/lib/fetcher.ts

```diff
@@ -0,0 +1,159 @@
+    const body: any = body_;
+    if (
+      typeof body === "object" &&
+      body !== null &&
+      typeof body.message === "string"
+    ) {
+      super(body.message);
+    } else {
+      super("Response error");
+    }
+
+    this.method = method;
+    this.body = body_;
+    this.status = response.status;
+    this.url = response.url;
+  }
+}
+
+// Error that is thrown by `Fetcher` methods when parsing the response
+// body fails.
+export class ResponseParseError extends Error {
+  public method: string;
+  public path: string;
+  public body: unknown;
+  public zodIssues: ZodIssue[];
+
+  public constructor(
+    method: string,
+    path: string,
+    body: unknown,
+    zodIssues: ZodIssue[],
+  ) {
+    super("Failed to parse response body");
+    this.method = method;
+    this.path = path;
+    this.body = body;
+    this.zodIssues = zodIssues;
+  }
+}
+
+export interface RequestOptions {
+  abort?: AbortSignal;
+}
+
+export interface FetchParams {
+  method: Method;
+  // Path to append to the `Fetcher`s base URL to get the final URL.
+  path: string;
+  // Object that is serialized into JSON and sent as the data.
+  body?: unknown;
+  // Query parameters to be serialized with URLSearchParams.
+  query?: Record<string, string | number | boolean>;
+  options?: RequestOptions;
+  headers?: Record<string, string>;
+}
+
+export type Method = "DELETE" | "GET" | "PATCH" | "POST" | "PUT";
+
+export class Fetcher {
+  #baseUrl: BaseUrl;
+
+  public constructor(baseUrl: BaseUrl) {
+    this.#baseUrl = baseUrl;
+  }
+
+  // Execute a fetch and parse the result with the provided schema.
+  // Return the parsed payload.
+  //
+  // Throws `ResponseError` if the response status code is not `200`.
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  public async fetchOk<T extends ZodType<any>>(
+    params: FetchParams,
+    schema: T,
+  ): Promise<TypeOf<T>> {
+    const response = await this.fetch(params);
+
+    const responseBody = await response.json();
+
+    if (!response.ok) {
+      throw new ResponseError(params.method, response, responseBody);
+    }
+
+    const result = schema.safeParse(responseBody);
+    if (result.success) {
+      return result.data;
+    } else {
+      throw new ResponseParseError(
+        params.method,
+        params.path,
+        responseBody,
+        result.error.errors,
+      );
+    }
+  }
+
+  // Execute a fetch and ignore the response body.
+  //
+  // Throws `ResponseError` if the response status code is not `200`.
+  public async fetchOkNoContent(params: FetchParams): Promise<void> {
+    const response = await this.fetch(params);
+
+    if (!response.ok) {
+      let responseBody = await response.text();
+      try {
+        responseBody = JSON.parse(responseBody);
+      } catch (_e: unknown) {
+        // We keep the original text response body.
+      }
+      throw new ResponseError(params.method, response, responseBody);
+    }
+  }
+
+  private async fetch({
```

Comment:
Hmm, if I understand the code correctly, it's a different problem. We're not really fetching the `rawPath` anywhere, we just supply it so that markdown can render some images. But I think it doesn't need to in all cases, like the `rawPath` prop should be optional and used depending on context. I don't think it makes sense to have it enabled for comments and issue/patch descriptions. It should only do image replacements for markdown files.

Reply:
I think we should look at it, but in a different PR.